pytorch_toolkit/formula_recognition/im2latex/data/utils.py

```diff
@@ -0,0 +1,469 @@
+
+    def __call__(self, imgs):
+        fx = np.random.uniform(self.scale_min, self.scale_max)
+        fy = fx  # np.random.uniform(fx, self.scale_max)
+        if not isinstance(imgs, list):
+            imgs = [imgs]
+        imgs = [cv.resize(img, dsize=None, fx=fx, fy=fy) for img in imgs]
+        return imgs
+
+
+class TransformRotate:
+    def __init__(self, angle):
+        self.angle = angle
+
+    def __repr__(self):
+        return "TransformRotate(angle={})".format(self.angle)
+
+    def __call__(self, imgs):
+        bound = self.angle
+        angle = np.random.uniform(low=0 - bound, high=bound)
+        if not isinstance(imgs, list):
+            imgs = [imgs]
+
+        img = imgs[0]
+        h, w = img.shape[0:2]
+        center = int(w / 2), int(h / 2)
+
+        M = cv.getRotationMatrix2D(center, angle=angle, scale=1)
+        src = np.array([[[0, 0]], [[w, 0]], [[0, h]], [[w, h]]])
+
+        new_pts = cv.transform(src, M)
+
+        bb = cv.boundingRect(new_pts)
+        empty_img_shape = (bb[3], bb[2])
+        shift_matrix = np.array([[0, 0, 0-bb[0]], [0, 0, 0 - bb[1]]])
+        M = M + shift_matrix
+        rotated = [cv.warpAffine(
+            img, M, empty_img_shape[::-1], borderMode=cv.BORDER_CONSTANT, borderValue=COLOR_WHITE) for img in imgs]
+
+        return rotated
+
+
+def get_timestamp():
+    return "{:%Y-%m-%d_%H-%M-%S}".format(datetime.datetime.now())
+
+
+def get_num_lines_in_file(path):
+    total = 0
+    with open(path) as f:
+        for line in f:
+            total += 1
+    return total
+
+
+def collate_fn(sign2id, batch, *, batch_transform=None):
+    # filter the pictures that have different width or height
+    size = batch[0]['img'].shape
+    batch = [img_formula for img_formula in batch
+             if img_formula['img'].shape == size]
+    # sort by the length of formula
+    # the purpose of the sort is to put the longest formula on the first place
+    # to get correct size of the tensor in the formulas2tensor function
+    batch.sort(key=lambda img_formula: len(img_formula['formula'].split()),
+               reverse=True)
+
+    imgs = [item['img'] for item in batch]
+    formulas = [item['formula'] for item in batch]
+    img_names = [item['img_name'] for item in batch]
+    formulas_tensor = formulas2tensor(formulas, sign2id)
+
+    if batch_transform:
+        imgs = batch_transform(imgs)
+    imgs = torch.stack(imgs, dim=0)
+
+    bsize = len(batch)
+    # Ground truth symbols that are used as a decoding step input for the next symbol prediction during training.
+    training_gt = torch.cat([torch.ones(bsize, 1).long()*START_TOKEN, formulas_tensor], dim=1)
+    # Ground truth values for the outputs of decoder. Used for loss computation.
+    loss_computation_gt = torch.cat([formulas_tensor, torch.ones(bsize, 1).long()*END_TOKEN], dim=1)
+    return img_names, imgs, training_gt, loss_computation_gt
+
+
+def create_list_of_transforms(transforms_list, ovino_ir=False):
```

Comment:
Vladimir, did you check this function in this repo on non-trivial use-cases?

Are you sure that this functions works in this repo correctly (and nothing is broken in complicated use cases)?

Reply:
The transformations themselves were tested, they are just wrappers on corresponding opencv transformations. As for this function (`create_list_of_transforms`), this constructor of transformation pipeline is straightforward and allows to construct any possible pipeline of the transformations and it is user's responsibility to construct meaningful pipeline, i.e. not to combine inconsistent transformations (e.g. Resize to (2, 2), though it is valid transformation).